feat(SBM): Implementato Stochastic Block Model

CristianCosci opened the pull request.

ncvescera committed:
SBM Link Reliability

Iniziato sviluppo SBM.
Iniziata implementazione calcolo Link Reliability.

Co-authored-by: F-A-B-R-I-Z-I-O <fabriziofagiolo18@gmail.com>
Co-authored-by: Cristian <CristianCosci@users.noreply.github.com>

ncvescera committed:
Implementato SBM

Completata implementazione del metodo SBM.

Co-authored-by: F-A-B-R-I-Z-I-O <fabriziofagiolo18@gmail.com>
Co-authored-by: Cristian <CristianCosci@users.noreply.github.com>

ncvescera committed:
Import Fix & refactoring

Aggiunto import nel relativo file init.
Refactoring del codice aggiungendo Type Hints,
piccole modifiche ai nomi delle variabili e linting.

CristianCosci committed:
Corretti piccoli typo su tipologie matrici e fissato seed

ncvescera commented on src/social_network_link_prediction/probabilistic_methods/sbm.py:
```diff
@@ -0,0 +1,144 @@
+
+
+def __random_changes(A_0: csr_matrix, p: float = .05, seed: int = 42) -> lil_matrix:
+    np.random.seed(seed)
```
Questo andrebbe messo all'inizio della funzione principale

```suggestion

```

ncvescera replied:
Risolto in 6dc47ae

ncvescera commented on src/social_network_link_prediction/probabilistic_methods/sbm.py:
```diff
@@ -0,0 +1,144 @@
+    return summing_result / np.sum(HP_mem)
+
+
+def stochastic_block_model(G: nx.Graph,
```
@CristianCosci si rompe con grafi con labels stringhe ?

ncvescera commented on src/social_network_link_prediction/probabilistic_methods/sbm.py:
```diff
@@ -0,0 +1,144 @@
+                           n: int,
+                           p: float = .05,
+                           seed: int = 42) -> csr_matrix:
+    A_0 = to_adjacency_matrix(G)
```
```suggestion
    np.random.seed(seed)
    A_0 = to_adjacency_matrix(G)
```

ncvescera replied:
Risolto in 6dc47ae

ncvescera committed:
fix(SBM): Risolto problema per grafi con stringhe

Ora il metodo funziona anche per grafi che hanno come etichette dei nodi
delle stringhe.
Migliorato utilizzo di np.random.seed()

ncvescera committed:
fix(SBM): Rimosso seed da funzioni

Rimosso il parametro seed dalle funzione che non ne necessitano più.
Riaggiunto il parametro seed ad una funzione che ne aveva bisogno per
passarlo come parametro alla funzione louvain_communities()
